### PR TITLE
Allowing keys with leading slashes for S3

### PR DESCRIPTION
**Pull request by spfink**
S3 allows keys to begin with slashes for unknown reasons. This change allows those keys to work with the SDK and our HTTP clients.

## Testing
Added integration test

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

## Checklist

- [x] I have read the **CONTRIBUTING** document
- [x] Local run of `mvn install` succeeds
- [x] My code follows the code style of this project
- [x] My change requires a change to the Javadoc documentation
- [x] I have updated the Javadoc documentation accordingly
- [x] I have read the **README** document
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed
- [ ] A short description of the change has been added to the **CHANGELOG**
- [ ] My change is to implement 1.11 parity feature and I have updated [LaunchChangelog](https://github.com/aws/aws-sdk-java-v2/blob/master/docs/LaunchChangelog.md)

## License

- [x] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

**Comment by millems** [comment]
Changelog?

File: core/protocols/aws-xml-protocol/src/main/java/software/amazon/awssdk/protocols/xml/internal/marshall/SimpleTypePathMarshaller.java [hunk 1]
```diff
@@ -37,7 +37,7 @@
      * so that it preserves the path structure.
      */
     public static final XmlMarshaller<String> GREEDY_STRING =
-        new SimplePathMarshaller<>(ValueToStringConverter.FROM_STRING, PathMarshaller.GREEDY);
+        new SimplePathMarshaller<>(ValueToStringConverter.FROM_STRING, PathMarshaller.GREEDY_WITH_SLASHES);
```
Comment (millems) [inline core/protocols/aws-xml-protocol/src/main/java/software/amazon/awssdk/protocols/xml/internal/marshall/SimpleTypePathMarshaller.java]: What services are affected by this change? How certain are we that they wouldn't be broken by this?

File: services/s3/src/it/java/software/amazon/awssdk/services/s3/KeysWithLeadingSlashIntegrationTest.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+import java.nio.charset.StandardCharsets;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import software.amazon.awssdk.core.sync.RequestBody;
+
+public class KeysWithLeadingSlashIntegrationTest extends S3IntegrationTestBase {
+
+    private static final String BUCKET = temporaryBucketName(KeysWithLeadingSlashIntegrationTest.class);
+    private static final String KEY = "/stupidkeywithillegalleadingslashthatsucks";
```
Comment (millems) [inline services/s3/src/it/java/software/amazon/awssdk/services/s3/KeysWithLeadingSlashIntegrationTest.java]: No passive aggression here. Nope.

File: core/protocols/aws-xml-protocol/src/main/java/software/amazon/awssdk/protocols/xml/internal/marshall/SimpleTypePathMarshaller.java (hunk 1 above)
Comment (spfink) [reply core/protocols/aws-xml-protocol/src/main/java/software/amazon/awssdk/protocols/xml/internal/marshall/SimpleTypePathMarshaller.java]: This only effects XML services of which none of them currently use the greedy path marshaller.

File: services/s3/src/it/java/software/amazon/awssdk/services/s3/KeysWithLeadingSlashIntegrationTest.java (hunk 2 above)
Comment (spfink) [reply services/s3/src/it/java/software/amazon/awssdk/services/s3/KeysWithLeadingSlashIntegrationTest.java]: Just a random key name I swear.

File: core/protocols/aws-xml-protocol/src/main/java/software/amazon/awssdk/protocols/xml/internal/marshall/SimpleTypePathMarshaller.java (hunk 1 above)
Comment (spfink) [reply core/protocols/aws-xml-protocol/src/main/java/software/amazon/awssdk/protocols/xml/internal/marshall/SimpleTypePathMarshaller.java]: Other than S3 that is.

**Comment by spfink** [comment]
I don't like to change the protocol tests as I've done here but I think in this case it's fine. 

I've confirmed no other XML service uses the greedy path besides S3 and this is the correct behavior for S3. We shouldn't see any new services use XML. It could effect existing services if they add operations with a greedy path in the future. At that point we could switch this to be an S3 only customization but that would introduce a significant amount of additional handwritten S3 code and a bunch of new codegen customizations.

**Pull request merged**